### PR TITLE
Batch check relations

### PR DESCRIPTION
**Pull request by patrickduffy95**
## Related issue(s)
https://github.com/ory/keto/issues/812

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments
NOTE: This is a "draft" PR that isn't complete and is meant to get early feedback on the proposed set of changes.

This change adds REST and gRPC endpoints for batch checking relations. The endpoint accepts a list of relation tuples to check, iterates through them, and returns a list of allowed responses.

**REST API**
New endpoint:
`POST /relation-tuples/batchCheck?max-depth=<depth>`
Request body:
```
{
    "tuples": [
        {
            "namespace": <namespace>,
            "object": <object>,
            "relation": <relation>,
            "subject_id": <subject_id>,
            "subject_set": <subject_set>
        }
    ]
}
```
Response:
```
{
    "results": [
        {
            "allowed": true
        }
    ]
}
```

**gRPC**
New RPC:
`CheckService/BatchCheck`
Request:
```
// The request for a CheckService.BatchCheck RPC.
// Checks a batch of relations.
message BatchCheckRequest {
  repeated RelationTuple tuples = 1;

  // This field is not implemented yet and has no effect.
  bool latest = 2;
  // This field is not implemented yet and has no effect.
  string snaptoken = 3;
  int32 max_depth = 4;
}
```
Response
```
// The response for a CheckService.Check rpc.
message BatchCheckResponse {
  //
  repeated CheckResponse results = 1;
}
```

Questions I'd like feedback on:
- Should we limit the number of tuples this batch API accepts?
- Should we parallelize the check requests?
  - A more optimized implementation that reduces the number of calls to the DB rather than iterating through the request tuples is probably more than I'd want to tackle
- Is implementing only the POST okay for the REST API?

### CODE REVIEW

**Comment by patrickduffy95** [comment]
@alnr would you be able to let me know if this PR is on the right track?

**Comment by aeneasr** [comment]
Hello, sorry for kot responding here. I think this feature is grand! @zepatrik and @hperl are finishing up some work on our end and can probably check out this PR in a week or two.

**Comment by PrimeDominus** [comment]
Thanks for implementing this. We've hit a roadblock and without this feature we have to write a bunch of custom code in our server which is getting cumbersome to maintain. Really looking forward to trying this.

**Comment by patrickduffy95** [comment]
> Hello, sorry for kot responding here. I think this feature is grand! @zepatrik and @hperl are finishing up some work on our end and can probably check out this PR in a week or two.

Thanks @aeneasr! Do you have a better idea now of when they might be able to take a look at it?

File: internal/check/handler.go [hunk 1]
```diff
@@ -329,3 +331,121 @@ func (h *Handler) Check(ctx context.Context, req *rts.CheckRequest) (*rts.CheckR
...
+			Allowed: allowed,
+		}
+	}
+
+	return results, nil
+}
+
+func (h *Handler) BatchCheck(ctx context.Context, req *rts.BatchCheckRequest) (*rts.BatchCheckResponse, error) {
+	responses := make([]*rts.CheckResponse, len(req.Tuples))
+	for i, reqTuple := range req.Tuples {
```
Comment (hperl) [inline internal/check/handler.go:432]: The logic is duplicated from the REST handler here. Ideally we'd implement this once and have gRPC and REST call that function.

**Review by hperl — CHANGES_REQUESTED** [review]
Thank you very much for your contribution! This looks very good already.

Besides the comments below, the main thing that is missing are some tests. I'd like to see tests for the happy path, going over the batch limit (bad request), graceful handling of on errornous tuple in the batch (should not fail the whole batch, but just that check), empty batch, and whatever edge case you can think of.

I left some comments below. As for your questions, here's my take on it:

> Should we limit the number of tuples this batch API accepts?

Yes, please make it configurable with a conservative default (10?)

> Should we parallelize the check requests?

Yes, I'd make the number of concurrent checks another parameter.

> Is implementing only the POST okay for the REST API?

Yes, that will suffice.

**Comment by patrickduffy95** [comment]
Thanks a lot for the review @hperl. I will let you know once this PR is updated.

File: internal/check/handler.go [hunk 2]
```diff
@@ -329,3 +331,121 @@ func (h *Handler) Check(ctx context.Context, req *rts.CheckRequest) (*rts.CheckR
...
+			results[i] = &CheckPermissionResult{
+				Allowed: false,
+			}
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		allowed, err := h.d.PermissionEngine().CheckIsMember(ctx, t[0], maxDepth)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (patrickduffy95) [reply internal/check/handler.go:420]: Adding more context to the error is a good idea. The question then is whether to return errors at the tuple level vs for the API as a whole. If there are situations where clients might expect an error and would want to treat an error the same as receiving a `false` allowed value, I could see a benefit in continuing on error and returning for each tuple if an error was generated. As far as I know though, that isn't a normal case, so I will just add tuple details to the error message.

[the remark this answers is not in the document]
Comment (patrickduffy95) [reply internal/check/handler.go:420]: Changed my mind and went for proceeding with checking the whole batch and adding the error to the response for each tuple

File: cmd/status/root_test.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ func TestStatusCmd(t *testing.T) {
 				defer cancel()
 
 				stdErr := cmdx.ExecExpectedErrCtx(ctx, t, newStatusCmd(), "--"+FlagEndpoint, string(serverType), "--"+ts.FlagRemote, ts.Addr[:len(ts.Addr)-1])
-				assert.Equal(t, "context deadline exceeded", stdErr)
+				assert.Contains(t, stdErr, "context deadline exceeded")
```
Comment (patrickduffy95) [inline cmd/status/root_test.go:40]: I think this started failing b/c of the dependency bump

**Comment by patrickduffy95** [comment]
@hperl I have update the PR based on your feedback and it is now ready for review.

**Comment by PrimeDominus** [comment]
I'm really looking forward to this functionality being one Ory Network. @hperl is there anything else that needs to be done before this can be merged?

File: internal/check/handler.go [hunk 4]
```diff
@@ -329,3 +354,155 @@ func (h *Handler) Check(ctx context.Context, req *rts.CheckRequest) (*rts.CheckR
...
+}
+
+// doBatchCheck is the HTTP entry point for checking batches of tuples
+func (h *Handler) doBatchCheck(ctx context.Context, body io.Reader, query url.Values) ([]*CheckPermissionResultWithError, error) {
+	maxDepth, err := x.GetMaxDepthFromQuery(query)
+	if err != nil {
+		return nil, err
+	}
+	parallelizationFactor := defaultBatchCheckParallelizationFactor
+	if query.Get(parallelizationFactorQueryParam) != "" {
```
Comment (hperl) [inline internal/check/handler.go:431]: Sorry for the misunderstanding, but I meant having this as a config parameter in the config (where you also added the max batch size). This is more a parameter the server admin sets to cap the max load of Keto than for a client to set, in my mind.

File: internal/check/engine.go [hunk 5]
```diff
@@ -264,3 +267,46 @@ func (e *Engine) astRelationFor(ctx context.Context, r *relationTuple) (*ast.Rel
...
+		return nil, errors.New("invalid parallelization factor")
+	}
+
+	wg := &sync.WaitGroup{}
+	sem := semaphore.NewWeighted(int64(parallelizationFactor))
+
+	mapper := e.d.ReadOnlyMapper()
+	results := make([]checkgroup.Result, len(tuples))
+	for outerIndex, outerTuple := range tuples {
+		sem.Acquire(context.Background(), 1) // Pass in background context to guarantee this won't return an error
```
[the remark this answers is not in the document]
Comment (patrickduffy95) [reply internal/check/engine.go]: Nice catch on the race condition. I've updated it to use `sync.errgroup.Group`

File: internal/check/handler.go (hunk 4 above)
Comment (patrickduffy95) [reply internal/check/handler.go]: No worries, I've updated it to be a config parameter

**Review by hperl — APPROVED** [review]
This looks perfect now! Thanks for all the good work! 🎉

**Comment by rbnbr** [comment]
Hey, thank you for the great feature!

Now with this being merged in the master branch, could we also extend the Rest HTTP API accordingly (https://www.ory.sh/docs/keto/reference/rest-api) and also make it available to the keto-client(s), especially for go (https://github.com/ory/keto-client-go)?

Thank you!

**Comment by ryukinix** [comment]
Very interesting! We were waiting for this since 2022. Awesome work!

**Comment by jcottongit** [comment]
@hperl Cheers for getting this in, can we get a docker image released with this feature included?
